File: epi_judge_java/epi/TreeRightSibling.java
package epi;

import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;
import epi.test_framework.TimedExecutor;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.List;
import java.util.Queue;

public class TreeRightSibling {
    public static void constructRightSibling(BinaryTreeNode<Integer> tree) {
        if (tree == null) return;
        Queue<BinaryTreeNode<Integer>> q = new ArrayDeque<>();
        q.offer(tree);

        while (!q.isEmpty()) {
            int size = q.size();
            BinaryTreeNode<Integer> prev = null;
            for (int i = 0; i < size; i++) {
                BinaryTreeNode<Integer> polled = q.poll();
                if (prev != null) {
                    prev.next = polled;
                }

                prev = polled;

                if (polled.left != null) q.offer(polled.left);
                if (polled.right != null) q.offer(polled.right);
            }
        }
    }

    public static class BinaryTreeNode<T> extends TreeLike<T, BinaryTreeNode<T>> {
        public T data;
        public BinaryTreeNode<T> left, right;
        public BinaryTreeNode<T> next = null; // Populates this field.

        public BinaryTreeNode(T data) {
            this.data = data;
        }

        @Override
        public T getData() {
            return data;
        }

        @Override
        public BinaryTreeNode<T> getLeft() {
            return left;
        }

        @Override
        public BinaryTreeNode<T> getRight() {
            return right;
        }
    }

    private static BinaryTreeNode<Integer> cloneTree(BinaryTree<Integer> original) {
        if (original == null) {
            return null;
        }
        BinaryTreeNode<Integer> cloned = new BinaryTreeNode<>(original.data);
        cloned.left = cloneTree(original.left);
        cloned.right = cloneTree(original.right);
        return cloned;
    }

    @EpiTest(testDataFile = "tree_right_sibling.tsv")
    public static List<List<Integer>>
    constructRightSiblingWrapper(TimedExecutor executor, BinaryTree<Integer> tree)
            throws Exception {
        BinaryTreeNode<Integer> cloned = cloneTree(tree);

        executor.run(() -> constructRightSibling(cloned));

        List<List<Integer>> result = new ArrayList<>();
        BinaryTreeNode<Integer> levelStart = cloned;
        while (levelStart != null) {
            List<Integer> level = new ArrayList<>();
            BinaryTreeNode<Integer> levelIt = levelStart;
            while (levelIt != null) {
                level.add(levelIt.data);
                levelIt = levelIt.next;
            }
            result.add(level);
            levelStart = levelStart.left;
        }
        return result;
    }

    public static void main(String[] args) {
        System.exit(
                GenericTest
                        .runFromAnnotations(args, "TreeRightSibling.java",
                                new Object() {
                                }.getClass().getEnclosingClass())
                        .ordinal());
    }
}

File: epi_judge_cpp/tree_right_sibling.cc
#include <memory>
#include <vector>

#include "test_framework/generic_test.h"
#include "test_framework/serialization_traits.h"
#include "test_framework/timed_executor.h"

using namespace std;
using test_framework::BinaryTreeSerializationTrait;

template<typename T>
struct BinaryTreeNode {
    T data;
    unique_ptr<BinaryTreeNode<T>> left, right;
    BinaryTreeNode<T> *next = nullptr;  // Populates this field.

    explicit BinaryTreeNode(T data) : data(data) {};
};

void ConstructRightSibling(BinaryTreeNode<int> *tree) {
    if (tree == nullptr) return;
    vector<BinaryTreeNode<int> *> level;
    level.push_back(tree);

    while (!level.empty()) {
        vector<BinaryTreeNode<int> *> temp;
        for (size_t i = 0; i < level.size(); i++) {
            auto curr = level[i];
            if (curr->left != nullptr) temp.push_back(curr->left.get());
            if (curr->right != nullptr) temp.push_back(curr->right.get());

            if (i + 1 < level.size()) {
                curr->next = level[i + 1];
            }
        }

        level = temp;
    }
}

namespace test_framework {
    template<>
    struct SerializationTrait<unique_ptr < BinaryTreeNode < int>>>
    : BinaryTreeSerializationTrait<unique_ptr < BinaryTreeNode < int>>, false> {
};
}  // namespace test_framework

std::vector<std::vector<int>> ConstructRightSiblingWrapper(
        TimedExecutor &executor, unique_ptr<BinaryTreeNode<int>> &tree) {
    executor.Run([&] { ConstructRightSibling(tree.get()); });

    std::vector<std::vector<int>> result;
    auto level_start = tree.get();
    while (level_start) {
        result.emplace_back();
        auto level_iter = level_start;
        while (level_iter) {
            result.back().push_back(level_iter->data);
            level_iter = level_iter->next;
        }
        level_start = level_start->left.get();
    }
    return result;
}

int main(int argc, char *argv[]) {
    std::vector<std::string> args{argv + 1, argv + argc};
    std::vector<std::string> param_names{"executor", "tree"};
    return GenericTestMain(
            args, "tree_right_sibling.cc", "tree_right_sibling.tsv",
            &ConstructRightSiblingWrapper, DefaultComparator{}, param_names);
}
